feat: introduce proposal design process (#3333)
deepset-ai/haystack: rfcs/README.md

to collect feedback from other project developers first, to make sure that the RFC is
is actually needed. As we're an open source community where everyone can impact the project, we all need to make an effort to build consensus.

When you're preparing for writing and submitting an RFC, talk the idea over on our official
[Discord server](https://haystack.deepset.ai/community/join) and in a Github
issue or discussion in the [Haystack repository](https://github.com/deepset-ai/haystack).

## The process

In short, to get a major feature added to Haystack, you usually first merge the RFC into
the RFC repo as a Markdown file. At that point, the RFC is 'active' and may be implemented with
the goal of eventually including it into the Haystack codebase.

To create an RFC:

1. Copy `0000-template.md` to `text/0000-my-feature.md`, where 'my-feature' is a descriptive name of the feature you're proposing. Don't assign an RFC number yet.
2. Fill in the RFC. Pay attention to details. RFCs that present convincing motivation, demonstrate an understanding of the feature impact, and honestly present the drawbacks and alternatives tend to be received well.

> 3. Submit a pull request. This ensures the RFC receives design feedback from a larger community. As the author, you should be prepared to revise it in response.

Reviewer: Instead of saying "Submit a pull request", can't we say something like "Submit the RFC with a pull request"? Saying "Submit a pull request" may mislead people to implement the RFC with the RFC md file. What do you think?

> 4. Rename the file using the PR number, for example from `text/0000-my-feature.md` to `text/4242-my-feature.md`.

Reviewer 2: Nitpick:
```suggestion
4. Rename the file using the RFC's own PR number, for example from `text/0000-my-feature.md` to `text/4242-my-feature.md`.
```

# Haystack RFCs

> Most of the changes to Haystack, including buf fixes and small improvements,

Reviewer 3: `buf` to `bug`

- A new feature that creates new API surface areas.
- A new component (Nodes, Pipelines, Document Stores).
- The removal of features that already shipped in the current minor version.
- A deep refactoring that would require new tests or introduce new dependencies
- A change complex enough that would require multiple steps to be delivered

Some changes do not require an RFC:

- Minor bug fixes
- Rephrasing, reorganizing, or otherwise "changing shape does not change meaning".
- Addition and removal of warnings or other error messages.
- Additions only likely to be noticed by other contributors, invisible to Haystack users.

In any case, the core developers might politely ask you to submit an RFC before merging
a new feature when they see fit.

## Before creating an RFC

A hastily-proposed RFC can hurt its chances of acceptance. Low quality proposals, proposals
for previously-rejected features, or those that don't fit into the near-term
[roadmap](https://github.com/orgs/deepset-ai/projects/3), may be rejected, which can be
demotivating for the new contributor. Laying some groundwork ahead of the RFC can make the
process smoother.

> Although there is no single way to prepare for submitting an RFC, it is generally a good idea

Author: On a second thought, let's not mention the roadmap and keep the instructions as simple as possible, we'll learn by doing here so I expect instructions to evolve anyways

- Submit a pull request. As a pull request the RFC will receive design feedback from the larger community, and the author should be prepared to revise it in response.
- Rename the file using the PR number, e.g. from `text/0000-my-feature.md` to `text/4242-my-feature.md`.
- Build consensus and integrate feedback. RFCs that have broad support are much more likely to make progress than those that don't receive any comments.
- Eventually, the core developers will decide whether the RFC is a candidate for inclusion in Haystack. Note that a team review may take a long time, and we suggest that you ask members of the community to review it first.
- RFCs that are candidates for inclusion in Haystack will enter a "final comment period" lasting 3 calendar days. The beginning of this period will be signaled with a comment and label on the RFCs pull request.
- An RFC can be modified based upon feedback from the core developers and community. Significant modifications may trigger a new final comment period.

> - An RFC may be rejected by the core developers after public discussion has settled and comments have been made summarizing the rationale for rejection. A core developer should then close the RFCs associated pull request.

Reviewer 4: ```suggestion
9. Core developers may reject an RFC once the public discussion and adding comments are over, adding the reason for rejection. A core developer then closes the PR related to the RFCs.
```

> - An RFC may be accepted at the close of its final comment period. A core developer will merge the RFCs associated pull request, at which point the RFC will become 'active'.

Reviewer 4: ```suggestion
10. Core developers may accept an RFC at the close of its final comment period. A core developer then merges the PR related to the RFCs. At this point, the RFC becomes 'active'. 
```

## The RFC lifecycle

> Once an RFC becomes active, then authors may implement it and submit the feature as one or more pull

Reviewer 4: ```suggestion
Once an RFC becomes active, the authors are free to implement it and submit the feature as one or more pull
```

> requests to the Haystack repo. Becoming 'active' is not a rubber stamp, and in particular still does

Reviewer 4: ```suggestion
requests to the Haystack repo. Becoming 'active' is not a rubber stamp, and in particular still doesn't
```

> not mean the feature will ultimately be merged; it does mean that the core team has agreed to it in

Reviewer 4: ```suggestion
mean the feature will ultimately be merged; it does mean that the core team has agreed to it in
```

> principle and are amenable to merging it if the implementation reflects the contents of the RFC.

Reviewer 4: ```suggestion
principle and is open to merging it if the implementation reflects the contents of the RFC.
```

> Furthermore, the fact that a given RFC has been accepted and is 'active' implies nothing about what

Reviewer 4: ```suggestion
The fact that a given RFC has been accepted and is 'active' doesn't imply it has a priority assigned or somebody's currently working on it. 
```

> priority is assigned to its implementation, nor whether anybody is currently working on it.

Reviewer 4: ```suggestion

```

> Modifications to active RFCs can be done in followup PRs. We strive to write each RFC in a manner that

Reviewer 4: ```suggestion
To change an active RFC, open follow-up PRs. Our goal is to write each RFC so that
```

> it will reflect the final design of the feature; but the nature of the process means that we cannot

Reviewer 4: ```suggestion
it reflects the final design of the feature, but the nature of the process means that we cannot
```

> expect every merged RFC to actually reflect what the end result will be at the time of the next release;

Reviewer 4: ```suggestion
expect every merged RFC to actually reflect what the end result will be at the time of the next release.
```

> therefore we try to keep each RFC document somewhat in sync with the feature as planned, tracking such

Reviewer 4: ```suggestion
That's why we try to keep each RFC document somewhat in sync with the feature as planned, tracking such
```

> changes via followup pull requests to the document.

Reviewer 4: ```suggestion
changes through follow-up pull requests to the document.
```

> The author of an RFC is not obligated to implement it. Of course, the RFC author (like any other developer)

Reviewer 4: ```suggestion
As the author of an RFC, you're not obligated to implement it. Of course, the RFC author (like any other developer)
```

is welcome to post an implementation for review after the RFC has been accepted.

## Inspiration

React's RFC process owes its inspiration to the [React](https://github.com/reactjs/rfcs) and

> [Rust](https://github.com/rust-lang/rfcs) RFC processes. We're open to change it if needed.

Reviewer 4: ```suggestion
[Rust](https://github.com/rust-lang/rfcs) RFC processes. We're open to changing it if needed.
```
